src/replication.c

```diff
@@ -197,27 +205,195 @@ void feedReplicationBacklogWithObject(robj *o) {
+ * update replica / replication-backlog referenced node and block position. */
+void feedReplicationBuffer(char *s, size_t len) {
+    static long long repl_block_id = 0;
+
+    if (server.repl_backlog == NULL) return;
+    server.master_repl_offset += len;
+    server.repl_backlog_histlen += len;
+
+    /* Install write handler for all replicas. */
+    prepareReplicasToWrite();
+
+    size_t start_pos = 0;
+    listNode *start_node = NULL;
+    int add_new_block = 0;
+    replBufBlock *used_last_block = NULL;
+    listNode *ln = listLast(server.repl_buffer_blocks);
+    replBufBlock *tail = ln? listNodeValue(ln): NULL;
+
+    /* Append to tail string when possible. */
+    if (tail && tail->size > tail->used) {
+        start_node = listLast(server.repl_buffer_blocks);
+        start_pos = tail->used;
+        used_last_block = tail;
+        /* Copy the part we can fit into the tail, and leave the rest for a
+         * new node */
+        size_t avail = tail->size - tail->used;
+        size_t copy = avail >= len? len: avail;
+        memcpy(tail->buf + tail->used, s, copy);
+        tail->used += copy;
+        s += copy;
+        len -= copy;
+    }
+    if (len) {
+        /* Create a new node, make sure it is allocated to at
+         * least PROTO_REPLY_CHUNK_BYTES */
+        size_t usable_size;
+        size_t size = len < PROTO_REPLY_CHUNK_BYTES? PROTO_REPLY_CHUNK_BYTES: len;
+        tail = zmalloc_usable(size + sizeof(replBufBlock), &usable_size);
+        /* Take over the allocation's internal fragmentation */
+        tail->size = usable_size - sizeof(replBufBlock);
+        tail->used = len;
+        tail->refcount = 0;
+        tail->repl_offset = server.master_repl_offset - tail->used + 1;
+        tail->id = repl_block_id++;
+        memcpy(tail->buf, s, len);
+        listAddNodeTail(server.repl_buffer_blocks, tail);
+        /* We also count the list node memory into replication buffer memroy. */
+        server.repl_buffer_mem += (usable_size + sizeof(listNode));
+        add_new_block = 1;
+        if (start_node == NULL) {
+            start_node = listLast(server.repl_buffer_blocks);
+            start_pos = 0;
+        }
+    }
+
+    /* For output buffer of replicas. */
+    listIter li;
+    listRewind(server.slaves,&li);
+    while((ln = listNext(&li))) {
+        client *slave = ln->value;
+        if (!canFeedReplicaReplBuffer(slave)) continue;
+
+        /* Update shared replication buffer start position. */
+        if (slave->ref_repl_buf_node == NULL) {
+            slave->ref_repl_buf_node = start_node;
+            slave->ref_block_pos = start_pos;
+            /* Only increase the start block reference count. */
+            ((replBufBlock *)listNodeValue(start_node))->refcount++;
+        }
+
+        /* Check output buffer limit only when add new block. */
+        if (add_new_block) closeClientOnOutputBufferLimitReached(slave, 1);
```

Comment:
they share memory, but that memory can still lead to issues and if a slow replica falls behind for too long, we wanna disconnect it and release the memory.
in that sense if there are two that fall behind, and both crossed the threshold, we'll still disconnect both and reclaim the memory.

also note (and maybe we can document that):
setting a replica output buffer limit that's lower than the backlog size is pointless since:
1. disconnecting the replica will not release any memory, but also:
2. even in the previous versions, such a setting would mean that if a replica managed to use that backlog for PSYNC, it'll immediately get disconnected.

so that was an always an invalid configuration, and we should document that.
@yoav-steinberg FYI (what we discussed yesterday).

so after we put that invalid configuration aside, we can look at this PR as a simple optimization.
i.e. semantically, when a replica uses less memory than the backlog limit config, we can argue that it uses the backlog and no memory of it's own.
and when replicas consume more memory than the backlog, they share memory, and also the backlog benefits from that as an extra size.

if we go with the above semantics, then my proposal for info fields makes sense, see https://github.com/redis/redis/pull/9166#discussion_r714043623

Reply:
> even in the previous versions, such a setting would mean that if a replica managed to use that backlog for PSYNC, it'll immediately get disconnected.

Yes, we always has this problem.

The season why I think we should keep `replication_backlog_size` and `client_output_buffer_limit for replicas` instead of only use `replication buffer` is that users could set them arbitrary, the network between master and replicas is generally fine,  so replicas output buffer is not always big. 

> after we put that invalid configuration aside, we can look at this PR as a simple optimization.

it seems an easy and effective way.

Reply:
p.s. we can't enforce one config to be greater than the other at config.c, because we don't know at which order they'll be changed.
what we can / should do, as Yoav suggested in another comment is that when we consider the replica-buffer limit, we `max()` it with the backlog size, so that setting it to a lower value will be meaningless.
plus a big comment in the code explaining why it's invalid (psync will succeed and then drop the replica), and maybe a mention of that fact in redis.conf.

Reply:
@ShooterIT can you implement the `max()` design suggested above?

Reply:
and maybe also add some documentation about the reason (both code and maybe redis.conf if you find a place for it)